sympy/matrices/eigen.py

```diff
@@ -19,6 +22,62 @@
 from .utilities import _iszero
 
 
+def _eigenvals_triangular(M, multiple=False):
+    """A fast decision for eigenvalues of an upper or a lower triangular
+    matrix.
+    """
+    diagonal_entries = [M[i, i] for i in range(M.rows)]
+    if multiple:
+        return diagonal_entries
+    return dict(Counter(diagonal_entries))
+
+
+def _eigenvals_eigenvects_mpmath(M):
+    prec = max([x._prec for x in M.atoms(Float)])
+
+    A = mp.matrix(M.evalf(n=prec_to_dps(prec)))
+    E, ER = mp.eig(A)
+    norm = mp.fsum([mp.re(e)**2 + mp.im(e)**2 for e in E])
+
+    while True:
+        prec_new = round(prec * 1.5)
```

Comment:
I would make this `prec_new = min(DEFAULT_MAXPREC, prec*1.5)` and then move the Exhausted test to after the norm check, making the condition `if prec_new == DEFAULT_MAXPREC`